### PR TITLE
Split transposed convolution to convTranspose2d

### PR DESCRIPTION
**Pull request by wangli69087**
@BruceDai Please help me review this PR. Thanks.

### CODE REVIEW

**Comment by BruceDai** [comment]
@wangli69087, please clean your commits, only submit the commit for "Split transposed convolution to conv transpose2d", thanks.

File: src/nn/graph_builder.ts [hunk 1]
```diff
@@ -367,6 +392,20 @@ export class MLGraphBuilder {
     return (new Conv2d(input, filter, options)).getFusedOutputs()[0];
   }
 
+  /**
+   * [spec](https://webmachinelearning.github.io/webnn/#dom-mlgraphbuilder-convtranspose2d)
+   */
+  convTranspose2d(input: MLOperand, filter: MLOperand, 
+    options: MLConvTranspose2dOptions = {}):
+      MLOperand {
```
Comment (BruceDai) [inline src/nn/graph_builder.ts:400]: Please update
```ts
  convTranspose2d(input: MLOperand, filter: MLOperand, 
    options: MLConvTranspose2dOptions = {}):
      MLOperand {
```
to
```ts
  convTranspose2d(
      input: MLOperand, filter: MLOperand,
      options: MLConvTranspose2dOptions = {}): MLOperand {
```

File: src/nn/graph_builder.ts [hunk 2]
```diff
@@ -75,13 +76,37 @@ export enum MLAutoPad {
...
  */
 export interface MLConv2dOptions {
+  padding?: [number, number, number, number];
+  strides?: [number, number];
+  dilations?: [number, number];
+  outputPadding?: [number, number];
+  autoPad?: MLAutoPad;
+  groups?: number;
+  inputLayout?: MLInputOperandLayout;
+  filterLayout?: MLFilterOperandLayout;
```
Comment (BruceDai) [inline src/nn/graph_builder.ts:86]: `MLFilterOperandLayout` is outdated, please change `MLFilterOperandLayout` to `MLConv2dFilterOperandLayout`, and also rename above enum `MLFilterOperandLayout` to `MLConv2dFilterOperandLayout`.

File: src/nn/graph_builder.ts [hunk 3]
```diff
@@ -75,13 +76,37 @@ export enum MLAutoPad {
...
   padding?: [number, number, number, number];
   strides?: [number, number];
   dilations?: [number, number];
   outputPadding?: [number, number];
   outputSizes?: [number, number];
   autoPad?: MLAutoPad;
-  transpose?: boolean;
   groups?: number;
   inputLayout?: MLInputOperandLayout;
   filterLayout?: MLFilterOperandLayout;
```
Comment (BruceDai) [inline src/nn/graph_builder.ts:112]: Please change `MLFilterOperandLayout` to `MLConvTranspose2dFilterOperandLayout`.

File: src/nn/ops/conv2d.ts [hunk 4]
```diff
@@ -22,8 +22,6 @@ export class Conv2d extends SingleOutputOperation implements FusedOperation {
   private filterLayout_?: MLFilterOperandLayout;
```
Comment (BruceDai) [inline src/nn/ops/conv2d.ts:22]: ditto

File: src/nn/ops/conv2d.ts [hunk 5]
```diff
@@ -44,19 +42,18 @@ export class Conv2d extends SingleOutputOperation implements FusedOperation {
...
-        options.bias, options.activation);
+        options.inputLayout, options.filterLayout, options.autoPad, 
+        options.outputPadding, options.bias, options.activation);
   }
 
   private initOptions(
       padding: [number, number, number, number] = [0, 0, 0, 0],
       strides: [number, number] = [1, 1], dilations: [number, number] = [1, 1],
       groups = 1, inputLayout: MLInputOperandLayout = MLInputOperandLayout.nchw,
       filterLayout: MLFilterOperandLayout = MLFilterOperandLayout.oihw,
```
Comment (BruceDai) [inline src/nn/ops/conv2d.ts:53]: ditto

File: src/nn/ops/conv2d.ts [hunk 6]
```diff
@@ -175,117 +156,80 @@ export class Conv2d extends SingleOutputOperation implements FusedOperation {
...
-      }
-      if (this.filter_ instanceof ConstantOperand) {
-        this.filterTensor_ = filter;
-        tf.keep(this.filterTensor_);
-      }
+  
+    // tf.conv2d filter layout (hwio): [filterHeight, filterWidth, inDepth,
+    // outDepth]
+    if (this.filterLayout_ === MLFilterOperandLayout.oihw) {
+      filter = tf.transpose(filter, [2, 3, 1, 0]);
```
Comment (BruceDai) [inline src/nn/ops/conv2d.ts:163]: ditto

File: src/nn/ops/convtranspose2d.ts [hunk 7]
```diff
@@ -0,0 +1,251 @@
+import * as tf from '@tensorflow/tfjs-core';
+import {ExplicitPadding} from '@tensorflow/tfjs-core/dist/ops/conv_util';
+
+import {MLAutoPad, MLConvTranspose2dOptions, MLFilterOperandLayout, MLInputOperandLayout} from '../graph_builder';
```
Comment (BruceDai) [inline src/nn/ops/convtranspose2d.ts:4]: ditto

File: src/nn/ops/convtranspose2d.ts [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+  implements FusedOperation {
+  private input_: MLOperand;
+  private filter_: MLOperand;
+  private bias_: MLOperand;
+  private padding_?: [number, number, number, number];
+  private strides_?: [number, number];
+  private dilations_?: [number, number];
+  private groups_?: number;
+  private inputLayout_?: MLInputOperandLayout;
+  private filterLayout_?: MLFilterOperandLayout;
```
Comment (BruceDai) [inline src/nn/ops/convtranspose2d.ts:23]: ditto

File: src/nn/ops/convtranspose2d.ts [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+        options.inputLayout, options.filterLayout, options.autoPad,
+        options.outputPadding, options.outputSizes,
+        options.bias, options.activation);
+  }
+
+  private initOptions(
+      padding: [number, number, number, number] = [0, 0, 0, 0],
+      strides: [number, number] = [1, 1], dilations: [number, number] = [1, 1],
+      groups = 1, inputLayout: MLInputOperandLayout = MLInputOperandLayout.nchw,
+      filterLayout: MLFilterOperandLayout = MLFilterOperandLayout.oihw,
```
Comment (BruceDai) [inline src/nn/ops/convtranspose2d.ts:57]: ditto

File: src/nn/ops/convtranspose2d.ts [hunk 10]
```diff
@@ -0,0 +1,251 @@
...
+    utils.assert(utils.isInteger(groups), 'The gourps parameter is invalid.');
+    this.groups_ = groups;
+
+    utils.assert(
+        inputLayout in MLInputOperandLayout,
+        'The input layout parameter is invalid.');
+    this.inputLayout_ = inputLayout;
+
+    utils.assert(
+        filterLayout in MLFilterOperandLayout,
```
Comment (BruceDai) [inline src/nn/ops/convtranspose2d.ts:87]: ditto

File: src/nn/ops/convtranspose2d.ts [hunk 11]
```diff
@@ -0,0 +1,251 @@
...
+    // tf.convTranspose2d input layout (nhwc): [batch, height, width, inDepth]
+    if (this.inputLayout_ === MLInputOperandLayout.nchw) {
+      // nchw -> nhwc
+      input = tf.transpose(input, [0, 2, 3, 1]);
+    }
+    if (this.filterTensor_ === undefined) {
+      filter = inputTensors.get(this.filter_) as tf.Tensor4D;
+      // tf.conv2dTranspose filter layout (hwoi): [filterHeight, filterWidth,
+      // outDepth, inDepth]
+      if (this.filterLayout_ === MLFilterOperandLayout.oihw) {
```
Comment (BruceDai) [inline src/nn/ops/convtranspose2d.ts:173]: ditto
Please process these tree convTranspose2d filter layout: "iohw" | "hwoi" | "ohwi".

File: test/ops/convtranspose2d.js [hunk 12]
```diff
@@ -0,0 +1,1102 @@
...
+    };
+    const options = {
+      transpose: true,
+      inputLayout: 'nchw',
+      filterLayout: 'ohwi',
+    };
+    testConvTranspose2d(input, filter, expected, options);
+  });
+
+  it('conv2d transpose nchw ihwo', function() {
```
Comment (BruceDai) [inline test/ops/convtranspose2d.js:122]: There's none `ihwo` ConvTranspose2d filter layout, please add some tests with `iohw` | `hwoi` filter layouts.

File: test/ops/convtranspose2d.js [hunk 13]
```diff
@@ -0,0 +1,1102 @@
...
+    };
+    const options = {
+      transpose: true,
+      inputLayout: 'nhwc',
+      filterLayout: 'oihw',
+    };
+    testConvTranspose2d(input, filter, expected, options);
+  });
+
+  it('conv2d transpose nhwc hwio', function() {
```
Comment (BruceDai) [inline test/ops/convtranspose2d.js:174]: ditto

**Review by BruceDai — CHANGES_REQUESTED** [review]
@wangli69087 Please update filter layout for conv2d and convTranspose2d ops, and add some tests to check convTranspose2d with `iohw `| `hwoi` filter layouts.

**Comment by BruceDai** [comment]
LGTM, thanks @wangli69087

**Review by BruceDai — APPROVED** [review]

File: src/nn/graph_builder.ts [hunk 14]
```diff
@@ -55,11 +56,11 @@ export interface MLClampOptions {
 /**
  * [spec](https://webmachinelearning.github.io/webnn/#enumdef-mlfilteroperandlayout)
```
Comment (BruceDai) [inline src/nn/graph_builder.ts:57]: @wangli69087 Please update this link, thanks.

File: src/nn/graph_builder.ts [hunk 15]
```diff
@@ -75,16 +76,40 @@ export enum MLAutoPad {
...
+  autoPad?: MLAutoPad;
+  groups?: number;
+  inputLayout?: MLInputOperandLayout;
+  filterLayout?: MLConv2dFilterOperandLayout;
+  bias?: MLOperand;
+  activation?: MLOperator;
+}
+
+/**
+ * [spec](https://webmachinelearning.github.io/webnn/#enumdef-mlfilteroperandlayout)
```
Comment (BruceDai) [inline src/nn/graph_builder.ts:92]: ditto

File: src/nn/graph_builder.ts [hunk 16]
```diff
@@ -75,16 +76,40 @@ export enum MLAutoPad {
  * [spec](https://webmachinelearning.github.io/webnn/#dictdef-mlconv2doptions)
  */
 export interface MLConv2dOptions {
+  padding?: [number, number, number, number];
+  strides?: [number, number];
+  dilations?: [number, number];
+  outputPadding?: [number, number];
```
Comment (BruceDai) [inline src/nn/graph_builder.ts:82]: There isn't `outputPadding` option for [`MLConv2dOptions`](https://webmachinelearning.github.io/webnn/#dictdef-mlconv2doptions), please remove it.

File: src/nn/ops/conv2d.ts [hunk 17]
```diff
@@ -37,26 +35,20 @@ export class Conv2d extends SingleOutputOperation implements FusedOperation {
...
       strides: [number, number] = [1, 1], dilations: [number, number] = [1, 1],
       groups = 1, inputLayout: MLInputOperandLayout = MLInputOperandLayout.nchw,
-      filterLayout: MLFilterOperandLayout = MLFilterOperandLayout.oihw,
-      autoPad: MLAutoPad = MLAutoPad.explicit, transpose = false,
-      outputPadding: [number, number] = [0, 0],
-      outputSizes: [number, number] = undefined, bias: MLOperand = undefined,
+      filterLayout: 
+      MLConv2dFilterOperandLayout = MLConv2dFilterOperandLayout.oihw,
+      autoPad: MLAutoPad = MLAutoPad.explicit, 
+      outputPadding: [number, number] = [0, 0], bias: MLOperand = undefined,
```
Comment (BruceDai) [inline src/nn/ops/conv2d.ts:51]: ditto

File: src/nn/ops/conv2d.ts [hunk 18]
```diff
@@ -19,11 +19,9 @@ export class Conv2d extends SingleOutputOperation implements FusedOperation {
   private dilations_?: [number, number];
   private groups_?: number;
   private inputLayout_?: MLInputOperandLayout;
-  private filterLayout_?: MLFilterOperandLayout;
+  private filterLayout_?: MLConv2dFilterOperandLayout;
   private autoPad_?: MLAutoPad;
   private outputPadding_?: [number, number];
```
Comment (BruceDai) [inline src/nn/ops/conv2d.ts:24]: Please remove this variable.

File: src/nn/ops/conv2d.ts [hunk 19]
```diff
@@ -37,26 +35,20 @@ export class Conv2d extends SingleOutputOperation implements FusedOperation {
...
-          options.padding === undefined),
-        'The padding parameter should be assigned when autoPad is explicit.');
-
     this.initOptions(
         options.padding, options.strides, options.dilations, options.groups,
-        options.inputLayout, options.filterLayout, options.autoPad,
-        options.transpose, options.outputPadding, options.outputSizes,
-        options.bias, options.activation);
+        options.inputLayout, options.filterLayout, options.autoPad, 
+        options.outputPadding, options.bias, options.activation);
```
Comment (BruceDai) [inline src/nn/ops/conv2d.ts:41]: ditto

File: src/nn/ops/conv2d.ts [hunk 20]
```diff
@@ -82,31 +74,15 @@ export class Conv2d extends SingleOutputOperation implements FusedOperation {
...
-          outputSizes === undefined ||
-              (utils.isIntegerArray(outputSizes) && outputSizes.length === 2),
-          'The outputSizes parameter is invalid.');
-      this.outputSizes_ = outputSizes;
-    } else {
-      this.outputPadding_ = [0, 0];
-      this.outputSizes_ = undefined;
-    }
-
+    this.outputPadding_ = outputPadding;
```
Comment (BruceDai) [inline src/nn/ops/conv2d.ts:84]: ditto

File: src/nn/ops/conv2d.ts [hunk 21]
```diff
@@ -175,117 +151,80 @@ export class Conv2d extends SingleOutputOperation implements FusedOperation {
...
+    }
+    if (this.filter_ instanceof ConstantOperand) {
+      this.filterTensor_ = filter;
+      tf.keep(this.filterTensor_);
+    }
     } else {
       filter = this.filterTensor_;
     }
     const padding: 'valid'|'same'|ExplicitPadding = utils.getPaddings(
         input, filter, this.padding_, this.strides_, this.outputPadding_,
```
Comment (BruceDai) [inline src/nn/ops/conv2d.ts:179]: Replace  `this.outputPadding` with `[0,0]`

File: test/ops/conv2d.js [hunk 22]
```diff
@@ -2688,1061 +2714,6 @@ describe('test conv2d', function() {
...
-    const options = {
-      strides: [3, 2],
-      outputPadding: [1, 1],
-      outputSizes: [10, 8],
-      transpose: true,
-    };
-    testConv2d(input, filter, expected, options);
-  });
-
   it('conv2d transpose false', function() {
```
Comment (BruceDai) [inline test/ops/conv2d.js:2717]: Move these tests to conv_transpose2d.js test file.

**Review by BruceDai — CHANGES_REQUESTED** [review]
@wangli69087 There are some links need to update, and please remove `outputPadding` option from `conv2d` op, move some tests to `conv_transpose2d.js` file.

File: src/nn/ops/conv_transpose2d.ts [hunk 23]
```diff
@@ -70,9 +70,13 @@ export class ConvTranspose2d extends SingleOutputOperation
         utils.isIntegerArray(dilations) && dilations.length === 2,
         'The dilations parameter is invalid.');
     this.dilations_ = dilations;
+    utils.assert(this.dilations_.every(v => v === 1), 
+      'The tf.convtranspose2d does not support dilations parameter.');
```
Comment (BruceDai) [inline src/nn/ops/conv_transpose2d.ts:74]: Please update `tf.convtranspose2d` to `tf.conv2dTranspose`.

File: src/nn/ops/conv_transpose2d.ts [hunk 24]
```diff
@@ -70,9 +70,13 @@ export class ConvTranspose2d extends SingleOutputOperation
...
 
     utils.assert(utils.isInteger(groups), 'The gourps parameter is invalid.');
     this.groups_ = groups;
+    utils.assert(this.groups_ === 1, 
+      'The tf.convtranspose2d does not support groups parameter.');
```
Comment (BruceDai) [inline src/nn/ops/conv_transpose2d.ts:79]: ditto

**Review by BruceDai — COMMENTED** [review]
@wangli69087 Pleas update error message for `groups` and `dilations` options checking.

**Review by BruceDai — APPROVED** [review]

**Comment by BruceDai** [comment]
LGTM, thanks @wangli69087
 
@huningxin Would you please take a look, thanks.

**Review by huningxin — APPROVED** [review]
LGTM, thanks!

**Pull request merged**